File: BOJ/1991.ts
import * as fs from 'fs';

const inputManager: Generator<string, any, unknown> = (function *() {
  const stdin = '/dev/stdin';
  const lines = fs.readFileSync(stdin).toString().split('\n');
  for (const line of lines) yield line;
  return null;
})();
const nextLine = (): string => inputManager.next().value;
const nextLineStrings = () => nextLine().split(' ');
let output = '';

type Node = {
  value: String;
  left: Node;
  right: Node;
} | null;

const ROOT_ALPHABET = 'A';
const NONE_ALPHABET = '.';

const printPreOrder = (node: Node) => {
  if (node === null) return;
  output += node.value;
  printPreOrder(node.left);
  printPreOrder(node.right);
};

const printInOrder = (node: Node) => {
  if (node === null) return;
  printInOrder(node.left);
  output += node.value;
  printInOrder(node.right);
};

const printPostOrder = (node: Node) => {
  if (node === null) return;
  printPostOrder(node.left);
  printPostOrder(node.right);
  output += node.value;
};

// get the number of nodes
const N = parseInt(nextLine());
const nodes = new Map<string, Node>();
nodes.set(NONE_ALPHABET, null);
for (let i = 0; i < N; i++) {
  const value = String.fromCharCode(65 + i);
  nodes.set(value, { value, left: null, right: null });
}

// construct the tree
for (let i = 0; i < N; i++) {
  const [n, l, r] = nextLineStrings();
  const node = nodes.get(n)!;
  const left = nodes.get(l)!;
  const right = nodes.get(r)!;
  node.left = left;
  node.right = right;
}

const root = nodes.get(ROOT_ALPHABET)!;
printPreOrder(root);
output += '\n';
printInOrder(root);
output += '\n';
printPostOrder(root);
output += '\n';

// print the answer
console.log(output);

File: BOJ/1991.cpp
#include <iostream>
#include <vector>

#define NONE_ALPHABET '.'
#define NONE_NUMBER -1

using namespace std;

struct Node {
	int left, right;
};

inline int alphabet_to_number(char c) {
  return c == NONE_ALPHABET ? NONE_NUMBER : c - 'A';
}

inline char number_to_alphabet(int n) {
  return n == NONE_NUMBER ? NONE_ALPHABET : 'A' + n;
}

void print_pre_order(vector<Node> &nodes, int root) {
	if (root == NONE_NUMBER) return;
	cout << number_to_alphabet(root);
	print_pre_order(nodes, nodes[root].left);
	print_pre_order(nodes, nodes[root].right);
}

void print_in_order(vector<Node> &nodes, int root) {
	if (root == NONE_NUMBER) return;
	print_in_order(nodes, nodes[root].left);
	cout << number_to_alphabet(root);
	print_in_order(nodes, nodes[root].right);
}

void print_post_order(vector<Node> &nodes, int root) {
	if (root == NONE_NUMBER) return;
	print_post_order(nodes, nodes[root].left);
	print_post_order(nodes, nodes[root].right);
	cout << number_to_alphabet(root);
}

int main() {
	int N;
	vector<Node> nodes;

	// get the number of nodes
	cin >> N;
	nodes.resize(N);

	// construct the tree
	while (N--) {
		char node, left, right;
		cin >> node >> left >> right;
		nodes[alphabet_to_number(node)].left = alphabet_to_number(left);
		nodes[alphabet_to_number(node)].right = alphabet_to_number(right);
	}

	print_pre_order(nodes, 0);
	cout << '\n';
	print_in_order(nodes, 0);
	cout << '\n';
	print_post_order(nodes, 0);
	cout << '\n';

	return 0;
}
